### PR TITLE
[Integration] Financial Modeling Prep Api Integration 

### PR DESCRIPTION
**Pull request by RiyanaD**
## Description

Implemented key features of the financial modeling prep api integration, including connecting to the financial modeling api, querying database and loading daily chart stock information for a specific symbol, and the limit feature which loads only a certain amount of data points. Created a read-me for these features and wrote tests to validate these basic functionalities.

Fixes https://github.com/mindsdb/mindsdb/issues/5525

## Type of change

- [ ] ⚡ New feature (non-breaking change which adds functionality)

## Verification Process

To ensure the changes are working as expected:

 - [ ]   Run tests: python -m unittest discover -s mindsdb/integrations/handlers/financial_modeling_prep_handler/tests
 
## Additional Media:

- [ ] [I have attached a brief loom video or screenshots showcasing the new functionality or change](https://www.loom.com/share/2cd6ebfe94a943518b88ae4a788e42ca?sid=b282475a-8822-4010-b763-b8f3c534d0ee).

## Checklist:

- [x] My code follows the style guidelines(PEP 8) of MindsDB.
- [x] I have appropriately commented on my code, especially in complex areas.
- [x] Necessary documentation updates are either made or tracked in issues.
- [x] Relevant unit and integration tests are updated or added.

### CODE REVIEW

File: mindsdb/integrations/handlers/financial_modeling_prep_handler/READ_ME.MD [hunk 1]
```diff
@@ -0,0 +1,80 @@
+# Financial Modeling Prep API Handler
```
Comment (ZoranPandovski) [inline mindsdb/integrations/handlers/financial_modeling_prep_handler/READ_ME.MD:1]: Let's name this as README.MD

File: .gitignore [hunk 2]
```diff
@@ -54,3 +54,8 @@ xgboost
 
 # Docs
 node_modules
+
+# Venv
+.venv/
+mindsdb/integrations/handlers/financial_modeling_prep_handler/commands.txt
```
Comment (ZoranPandovski) [inline .gitignore:60]: Let's remove this two lines, we don't need them in ignore

File: mindsdb/integrations/handlers/financial_modeling_prep_handler/financial_modeling_handler.py [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+    HandlerStatusResponse as StatusResponse,
+    HandlerResponse as Response,
+    RESPONSE_TYPE
+)
+from mindsdb.integrations.handlers.financial_modeling_prep_handler.financial_modeling_tables import FinancialModelingTradesTable
+
+from urllib.request import urlopen
+from mindsdb.utilities import log
+#from mindsdb_sql import parse_sql
+import certifi
```
Comment (ZoranPandovski) [inline mindsdb/integrations/handlers/financial_modeling_prep_handler/financial_modeling_handler.py:15]: This is imported but not used?

**Review by ZoranPandovski — CHANGES_REQUESTED** [review]
Few minor commennts

File: mindsdb/integrations/handlers/financial_modeling_prep_handler/financial_modeling_handler.py [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+    def connect(self) -> StatusResponse: 
+        self.is_connected = True
+        return StatusResponse(success = True)
+    
+    def check_connection(self) -> StatusResponse:
+        """ Check connection to the handler
+        Returns:
+            HandlerStatusResponse
+        """
+        return StatusResponse(success = True)
```
Comment (MinuraPunchihewa) [inline mindsdb/integrations/handlers/financial_modeling_prep_handler/financial_modeling_handler.py:49]: In the check_connection method, I suggest we check the validity of the API key that has been provided.

File: mindsdb/integrations/handlers/financial_modeling_prep_handler/financial_modeling_handler.py [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+                "FINANCIAL_MODELING engine requires an API key. Retrieve an API key from https://site.financialmodelingprep.com/developer. See financial_modeling_prep_handler/README.MD on how to include API key in query."
+            )
+        self.api_key = connection_data['api_key']
+        self.client = None
+        self.is_connected = False
+
+        daily_chart_table = FinancialModelingTradesTable(self) 
+        self._register_table('daily_chart_table', daily_chart_table)
+
+    def connect(self) -> StatusResponse:
```
Comment (MinuraPunchihewa) [inline mindsdb/integrations/handlers/financial_modeling_prep_handler/financial_modeling_handler.py:40]: The `connect()` method of our API handlers is typically expected to return an instance of a client (package) that will allow us to interact with the API in question. Not an instance of `StatusResponse` (although this is incorrectly used as a type hint in some of our handlers). [Here](https://github.com/mindsdb/mindsdb/blob/bc102f4011837d289bca43d7978717f408818e6b/mindsdb/integrations/handlers/ms_teams_handler/ms_teams_handler.py#L55) is a good example of a correct implementation. Now, I understand that since you are using `requests` here, there is no client instance as such that can be returned. Therefore, you can choose to return something else; for instance the `base_url` and then you can use this within your `BaseResultsTable` like `base_url = self.handler.connect()`. It is completely up to you how you want to handle this, but in essence, what needs to be done is to remove the `StatusResponse` logic from `connect()`.

File: mindsdb/integrations/handlers/financial_modeling_prep_handler/financial_modeling_handler.py [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+        return StatusResponse(success = True)
+    
+    def check_connection(self) -> StatusResponse:
+        """ Check connection to the handler
+        Returns:
+            HandlerStatusResponse
+        """
+        return StatusResponse(success = True)
+
+    def get_daily_chart(self, params: Dict = None) -> pd.DataFrame:
```
Comment (MinuraPunchihewa) [inline mindsdb/integrations/handlers/financial_modeling_prep_handler/financial_modeling_handler.py:51]: Let's move this method to the `FinancialModelingTradesTable` class; since the logic belongs there, I think it is a good idea to do so.

File: mindsdb/integrations/handlers/financial_modeling_prep_handler/financial_modeling_handler.py [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+            )
+        )
+
+        if historical:
+            return pd.DataFrame(historical)
+        else:
+            return pd.DataFrame() 
+
+
+    def call_financial_modeling_api(self, endpoint_name: str = None, params: Dict = None) -> pd.DataFrame:
```
Comment (MinuraPunchihewa) [inline mindsdb/integrations/handlers/financial_modeling_prep_handler/financial_modeling_handler.py:90]: I think we can remove this method once we move the `get_daily_chart` method. If we are to add more tables, I think it is better to add the relevant logic (and its methods) to those tables.

File: mindsdb/integrations/handlers/financial_modeling_prep_handler/financial_modeling_handler.py [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+        if "api_key" not in connection_data:
+            raise Exception(
+                "FINANCIAL_MODELING engine requires an API key. Retrieve an API key from https://site.financialmodelingprep.com/developer. See financial_modeling_prep_handler/README.MD on how to include API key in query."
+            )
+        self.api_key = connection_data['api_key']
+        self.client = None
+        self.is_connected = False
+
+        daily_chart_table = FinancialModelingTradesTable(self) 
+        self._register_table('daily_chart_table', daily_chart_table)
```
Comment (MinuraPunchihewa) [inline mindsdb/integrations/handlers/financial_modeling_prep_handler/financial_modeling_handler.py:38]: I think we should give a more meaningful name to this table; maybe `historial_price` is more accurate? Let me know if you have a better idea. You can also rename the name of the class (`FinancialModelingTradesTable`) with this change, just to make things more consistent.

**Review by MinuraPunchihewa — CHANGES_REQUESTED** [review]

File: mindsdb/integrations/handlers/financial_modeling_prep_handler/financial_modeling_handler.py (hunk 5 above)
Comment (RiyanaD) [reply mindsdb/integrations/handlers/financial_modeling_prep_handler/financial_modeling_handler.py]: Makes sense! I adjusted the connect() method to return the base_url instead of a StatusResponse instance.

File: mindsdb/integrations/handlers/financial_modeling_prep_handler/financial_modeling_handler.py (hunk 8 above)
Comment (RiyanaD) [reply mindsdb/integrations/handlers/financial_modeling_prep_handler/financial_modeling_handler.py]: I've updated the table and class names to `historial_price` and `HistoricalPriceTradesTable` to reflect the purpose of this table more accurately!

File: mindsdb/integrations/handlers/financial_modeling_prep_handler/financial_modeling_handler.py (hunk 4 above)
Comment (RiyanaD) [reply mindsdb/integrations/handlers/financial_modeling_prep_handler/financial_modeling_handler.py]: I added some code to send an api request to a financial modeling endpoint and check if the status response is 200, if not I raise an exception.

File: mindsdb/integrations/handlers/databend_handler/tests/__init__.py
Comment (MinuraPunchihewa) [inline mindsdb/integrations/handlers/databend_handler/tests/__init__.py:1]: Why has this file been removed? Can you revert this change?

File: .gitignore [hunk 9]
```diff
@@ -54,3 +54,6 @@ xgboost
 
 # Docs
 node_modules
+
+mindsdb/share/man/man1/isympy.1
```
Comment (MinuraPunchihewa) [inline .gitignore:58]: These entries in the .gitignore do not seem to be general, but rather specific to your work environment. Let's remove them.

File: mindsdb/integrations/handlers/financial_modeling_prep_handler/__init__.py [hunk 10]
```diff
@@ -0,0 +1,19 @@
...
+from .__about__ import __version__ as version, __description__ as description
+try:
+    from .financial_modeling_handler import FinancialModelingHandler as Handler
+    import_error = None
+except Exception as e:
+    Handler = None
+    import_error = e
+
+title = 'Financial_Modeling'
+name = "financial_modeling"
```
Comment (MinuraPunchihewa) [inline mindsdb/integrations/handlers/financial_modeling_prep_handler/__init__.py:12]: I think we should change the name of the integration to `financial_modelling_prep`. financial_modelling seems a little too generic. Let's change the title, example queries and the other aspects of the code to reflect this as well.

File: mindsdb/integrations/handlers/financial_modeling_prep_handler/README.MD [hunk 11]
```diff
@@ -0,0 +1,80 @@
+# Financial Modeling Prep API Handler
```
Comment (MinuraPunchihewa) [inline mindsdb/integrations/handlers/financial_modeling_prep_handler/README.MD:1]: The queries in your examples use a table called `daily_chart_table`, but this does not seem to exist.

File: mindsdb/integrations/handlers/financial_modeling_prep_handler/financial_modeling_handler.py [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+        if "api_key" not in connection_data:
+            raise Exception(
+                "FINANCIAL_MODELING engine requires an API key. Retrieve an API key from https://site.financialmodelingprep.com/developer. See financial_modeling_prep_handler/README.MD on how to include API key in query."
+            )
+        self.api_key = connection_data['api_key']
+        self.client = None
+        self.is_connected = False
+
+        historical_price_table = HistoricalPriceTable(self) 
+        self._register_table('historical_price_table', historical_price_table)
```
Comment (MinuraPunchihewa) [inline mindsdb/integrations/handlers/financial_modeling_prep_handler/financial_modeling_handler.py:38]: Can we give this table a more meaningful name? We should definitely remove `_table` from the name. Maybe `historial_prices` or something else you think is more appropriate?

File: mindsdb/integrations/handlers/financial_modeling_prep_handler/financial_modeling_handler.py [hunk 13]
```diff
@@ -0,0 +1,61 @@
...
+
+from urllib.request import urlopen
+from mindsdb.utilities import log
+import json
+import requests
+_FINANCIAL_MODELING_URL = 'https://financialmodelingprep.com/api/v3/'
+
+logger = log.getLogger(__name__)
+
+class FinancialModelingHandler(APIHandler):
```
Comment (MinuraPunchihewa) [inline mindsdb/integrations/handlers/financial_modeling_prep_handler/financial_modeling_handler.py:20]: I believe you will need to add a `native_query()` method to your handler as well.

**Review by MinuraPunchihewa — CHANGES_REQUESTED** [review]
I have left a few comments that need to be addressed. Please note that there a few lint issues that need to fixed as well. You can find them here: https://github.com/mindsdb/mindsdb/actions/runs/10927860432/job/30851504800?pr=9427

File: mindsdb/integrations/handlers/financial_modeling_prep_handler/financial_modeling_handler.py (hunk 13 above)
Comment (RiyanaD) [reply mindsdb/integrations/handlers/financial_modeling_prep_handler/financial_modeling_handler.py:20]: I believe the native_query method is used when the API can take in raw SQL queries, however the financial modeling prep API doesn't. What should the native_query() method consist of in this case?

Comment (MinuraPunchihewa) [reply mindsdb/integrations/handlers/financial_modeling_prep_handler/financial_modeling_handler.py:18]: Yes, you are correct! Sorry, I was mistaken.

**Review by MinuraPunchihewa — COMMENTED** [review]
Hey @RiyanaD,
This looks good now. I left some minor comments that are all related to the wording in the README. You also need to revert the changes you've made to `.gitignore` and the removal of the `__init__.py` file in the tests for the Databend handler.

We should be ready to merge then!

Oh, you also need to resolve all of the lint issues given here: https://github.com/mindsdb/mindsdb/actions/runs/11227328900/job/31394147546?pr=9427

**Comment by mindsdbadmin** [comment]
<br/>Thank you for your submission, we really appreciate it. Like many open-source projects, we ask that you sign our [Contributor License Agreement](https://github.com/mindsdb/mindsdb/blob/main/assets/contributions-agreement/individual-contributor.md) before we can accept your contribution. You can sign the CLA by just posting a Pull Request Comment same as the below format.<br/>
   - - -
   I have read the CLA Document and I hereby sign the CLA
   - - -
   <sub>You can retrigger this bot by commenting **recheck** in this Pull Request. </sub><sub>Posted by the **CLA Assistant Lite bot**.</sub>

File: mindsdb/integrations/handlers/databend_handler/tests/__init__.py
Comment (RiyanaD) [reply mindsdb/integrations/handlers/databend_handler/tests/__init__.py:1]: I used 'git checkout 2f031fe9f56e9302c6e2277c164cf74dba494690 -- mindsdb/integrations/handlers/databend_handler/tests/__init__.py' to revert the file back to its state before it was removed, so it is now back in my directory.

**Comment by RiyanaD** [comment]
I have read the CLA Document and I hereby sign the CLA.

File: mindsdb/integrations/handlers/financial_modeling_prep_handler/financial_modeling_handler.py [hunk 14]
```diff
@@ -0,0 +1,57 @@
...
+from typing import Dict
+from mindsdb.integrations.libs.api_handler import APIHandler
+from mindsdb.integrations.libs.response import (
+    HandlerStatusResponse as StatusResponse,
+)
+from mindsdb.integrations.handlers.financial_modeling_prep_handler.financial_modeling_tables import HistoricalPriceTable
+
+from mindsdb.utilities import log
+import requests
+_FINANCIAL_MODELING_URL = 'https://financialmodelingprep.com/api/v3/'
```
Comment (MinuraPunchihewa) [inline mindsdb/integrations/handlers/financial_modeling_prep_handler/financial_modeling_handler.py:11]: This variable does not seem to be in use anywhere. Let's either use it in the code (within the `connect()` method) or remove it completely. I prefer the first option because that will allow us to change it easily in the future if required. Let's also give it a more meaningful name, maybe `FMP_BASE_URL`?

**Review by MinuraPunchihewa — CHANGES_REQUESTED** [review]
Hi Riyana,
I left one last comment (I think this will be my last), but you still have a bunch of [lint issues](https://github.com/mindsdb/mindsdb/actions/runs/11434598401/job/31808447889?pr=9427) that need to be resolved/

**Comment by RiyanaD** [comment]
> Hi Riyana, I left one last comment (I think this will be my last), but you still have a bunch of [lint issues](https://github.com/mindsdb/mindsdb/actions/runs/11434598401/job/31808447889?pr=9427) that need to be resolved/

The line numbers changed while I was editing the files which caused some lint issues to be missed, but everything should be fixed now!

**Review by MinuraPunchihewa — COMMENTED** [review]
Hey @RiyanaD,
Unfortunately, there is one last [lint issue](https://github.com/mindsdb/mindsdb/actions/runs/11466401010/job/31928064825?pr=9427) to be solved. I think you just need to add a blank newline to the end of your handler module.

**Comment by RiyanaD** [comment]
> Hey @RiyanaD, Unfortunately, there is one last [lint issue](https://github.com/mindsdb/mindsdb/actions/runs/11466401010/job/31928064825?pr=9427) to be solved. I think you just need to add a blank newline to the end of your handler module.

Fixed.

**Review by MinuraPunchihewa — APPROVED** [review]

**Review by ZoranPandovski — APPROVED** [review]
